### PR TITLE
Add support to queue songs with s!stream

### PR DESCRIPTION
**Pull request by dongtandung2001**

### CODE REVIEW

File: src/commands/util/stream.js [hunk 1]
```diff
@@ -24,43 +24,41 @@ const ytdl = require('ytdl-core-discord');
...
-};
 // audio object
 let audio = {
   queue: [],
   player: createAudioPlayer(),
 };
 
+// get next audio resource to play
+// create new AudioResource for audio to play
+const getNextResource = async (url) => {
```
Comment (evanugarte) [inline src/commands/util/stream.js:35]: instead of having this function, can we just call createAudioResource directly?

**Review by evanugarte — COMMENTED** [review]
lets change audio to look like
```js
let audio = {
  upcoming: [],
  history: [],
  player: createAudioPlayer(),
};
```
1. `upcoming` will hold an array of youtube urls that have not been played yet, where index 0 is the next video we wish to play
2. `history` will hold an array of youtube urls that are currently playing/been played, where index 0 is the earliest video that was played in the queue

we have two callback functions for `AudioPlayerStatus.Idle` and `AudioPlayerStatus.Playing`.

1. `AudioPlayerStatus.Idle` will try to pop the leftmost item from `upcoming`, for example:
```js
const latestItem = audio.upcoming.shift();
```
if `latestItem` is not `undefined`, we push it to `audio.history` and then do
```js
audio.player.play(await getNextResource(latestItem));
```
if the array is empty, `latestItem` will be `undefined`

1. `AudioPlayerStatus.Playing` will look at `audio.history[audio.history.length - 1]` and use that url for the `Now playing` message. it will not modify any array.


in the `execute` callback, when the bot recieves url, it will push the url to `audio.upcoming` if `audio.player.state.status === AudioPlayerStatus.Playing`, else it will push the url to `audio.history` and play the given url.

File: src/commands/util/stream.js [hunk 2]
```diff
@@ -92,24 +98,32 @@ module.exports = new Command({
       }
       // check if url is valid
       // would be better if can check playable url
-      if (isValidUrl(url)) {
-        try {
+      if (ytdl.validateURL(url)) {
+        // audio.queue.push(url);
```
Comment (evanugarte) [inline src/commands/util/stream.js:102]:
```suggestion
```

File: src/commands/util/stream.js [hunk 3]
```diff
@@ -92,24 +98,32 @@ module.exports = new Command({
...
+        const { videoDetails } = await ytdl.getInfo(url);
+        if (audio.player.state.status === AudioPlayerStatus.Playing) {
+          audio.upcoming.push(url);
+          message.reply(`Added track \`${videoDetails.title}\``);
+        }
+        else {
```
Comment (evanugarte) [inline src/commands/util/stream.js:108]:
```suggestion
        } else {
```

File: src/commands/util/stream.js [hunk 4]
```diff
@@ -92,24 +98,32 @@ module.exports = new Command({
...
       else {
         if (args[0] === undefined)
           message.reply(`Usage: 
           \`${prefix}stream <url>: Play a track\``);
         else {
-          if (args[0] === 'stop') {
+          if (args[0] === 'skip') {
+            audio.player.stop();
+          }
+          else if (args[0] === 'stop') {
```
Comment (evanugarte) [inline src/commands/util/stream.js:123]:
```suggestion
          } else if (args[0] === 'stop') {
```

File: src/commands/util/stream.js [hunk 5]
```diff
@@ -92,24 +98,32 @@ module.exports = new Command({
...
         }
       }
       else {
         if (args[0] === undefined)
           message.reply(`Usage: 
           \`${prefix}stream <url>: Play a track\``);
         else {
-          if (args[0] === 'stop') {
+          if (args[0] === 'skip') {
+            audio.player.stop();
```
Comment (evanugarte) [inline src/commands/util/stream.js:121]: did you test this? where doing s!stream skip stops and plays the next song

File: src/commands/util/stream.js [hunk 6]
```diff
@@ -24,43 +24,49 @@ const ytdl = require('ytdl-core-discord');
...
+  }
+
+});
+
+// playing state
+// handle Playing state
+audio.player.on(AudioPlayerStatus.Playing, async () => {
+
+  // get music's info
+  const { videoDetails: jsonData } = await ytdl.getInfo(
```
Comment (evanugarte) [inline src/commands/util/stream.js:66]:
```suggestion
audio.player.on(AudioPlayerStatus.Playing, async () => {
  const { videoDetails: jsonData } = await ytdl.getInfo(
```

File: src/commands/util/stream.js (hunk 5 above)
Comment (dongtandung2001) [reply src/commands/util/stream.js:120]: yes, i've already tested it

**Review by evanugarte — APPROVED** [review]

**Pull request merged**